### DISCUSSION TITLE
Help debug tailwind css not being applied only in production (Rails 8.0 and ActiveAdmin 4.0.0.beta15)

### DISCUSSION THREAD

**Discussion by yang-gojuly** [activeadmin/activeadmin#8585] [Help]
Hi there, 

This is related to this post I made here (https://github.com/activeadmin/activeadmin/discussions/8583) and somewhat related to (https://github.com/activeadmin/activeadmin/discussions/8538). 

In development, everything looks beautiful and works great like the following. 

<img width="1501" alt="Screenshot 2024-12-23 at 3 56 28 PM" src="https://github.com/user-attachments/assets/f1a3e01c-6b75-46bc-bb80-518b567ef69c" />

But in production, it looks like none of CSS is being applied. 

<img width="1507" alt="Screenshot 2024-12-23 at 3 57 20 PM" src="https://github.com/user-attachments/assets/261e5311-84bd-444a-888c-ea57f6138dcf" />

My setup isn't too different from @PoroshkinaVV's in https://github.com/activeadmin/activeadmin/discussions/8538...., and can't figure out how to go about fixing it. I have a react app that also uses tailwind css, and it looks fine in both development and production environments.

My environment: 

```
ruby 3.3.5
node 22.9.0
yarn 1.22.22
```

Gems:

```
rails 8.0.0
importmap-rails 2.0.3
cssbundling-rails 1.4.1
propshaft 1.1.0
```

package.json

```
{
  "name": "julyai",
  "private": "true",
  "engines": {
    "node": "22.x"
  },
  "dependencies": {
    "@activeadmin/activeadmin": "4.0.0-beta15",
    "postcss": "^8.4.49",
    "tailwindcss": "^3.4.17",
    ....
   },
  "scripts": {
    "build:js": "node esbuild.config.js",
    "build:css": "tailwindcss -i ./app/assets/stylesheets/application.css -o ./app/assets/builds/application.css --minify && tailwindcss -i ./app/assets/stylesheets/active_admin.css -o ./app/assets/builds/active_admin.css --minify -c tailwind-active_admin.config.js",
    "build": "npm-run-all --parallel build:*"
  },
}
```

In `application.html.slim`,
```
    = javascript_include_tag "application", "data-turbo-track": "reload", defer: true
    = stylesheet_link_tag "application", "data-turbo-track": "reload"
```

I did a different version of this to use `tailwindcss-rails` gem just to see if it would make a difference (by moving `tailwind.config.js` to `/config` and removing `tailwindcss -i ./app/assets/stylesheets/application.css ....` from `build:css`) , but the result was the same. 

Any help in this matter would be very much appreciated. 

Thank you,

**Comment by yang-gojuly** [comment]
Anyone, please???  Pretty please?

**Reply by PoroshkinaVV** [reply]
@yang-gojuly hello. Want to help you, but I have not run my application in production yet. Run it locally with production mode, and have the same issue. Later I'll try to fix it

**Reply by PoroshkinaVV** [reply]
I run this command.
```
RAILS_ENV=production rails assets:precompile
```
Get an error, that said I need to add `gem 'tailwindcss-ruby'`. Did it. Run command again. And then run 
```
RAILS_ENV=production rails s
```

That's worked for me. So I think in your deployment script you must add step with precompiling assets.

**Reply by yang-gojuly** [reply]
Hi @PoroshkinaVV, 

Thank you so much for your attempt. I really appreciate it. But, it didn't work for me. My rails is used as both back-end API server and serving javascript files for react front-end. I was also using `tailwindcss` installed via yarn, not via `tailwind-rails` gem. 

So, both js and css building were done via yarn like the follwing: 
```
  "scripts": {
    "build:js": "node esbuild.config.js",
    "build:css": "tailwindcss -i ./app/assets/stylesheets/application.css -o ./app/assets/builds/application.css --minify",
    "build:cssAdmin": "tailwindcss -i ./app/assets/stylesheets/active_admin.css -o ./app/assets/builds/active_admin.css --minify -c tailwind-active_admin.config.js",
    "build": "npm-run-all --parallel build:*"
  },
```

And since adding `tailwindcss-ruby` worked for you, I removed `tailwindcss` via yarn and installed `tailwind-rails` gem (which also installs `tailwindcss-ruby`) to see if that'd work. Since `tailwindcss:build` is automatically attached to `assets:precompile`, build scripts changed to the following:
```
  "scripts": {
    "build": "node esbuild.config.js",
    "build:css": "tailwindcss -i ./app/assets/stylesheets/active_admin.css -o ./app/assets/builds/active_admin.css --minify -c tailwind-active_admin.config.js"
  },
```

Deployment works fine, react app looks fine, but activeadmin still looks like only half of CSS is applied like this: 
<img width="1495" alt="Screenshot 2025-01-18 at 12 04 51 PM" src="https://github.com/user-attachments/assets/d00fc11c-a627-435e-addd-01e90f9190a7" />

It's quite driving me bonkers at this point......

Thank you so much for trying to help. Again, I really appreciate it.

**Comment by jsntv200** [comment]
Just ran into this issue and the problem is `cssbundling-rails` will only ever run `build:css` and `tailwindcss-rails` will only build from one source. The simplest fix I found is to add a rake task to build the admin css, this should work for your code:

```json
"scripts": {
  "build": "node esbuild.config.js",
  "build:css": "tailwindcss -i ./app/assets/stylesheets/application.css -o ./app/assets/builds/application.css --minify",
  "build:cssAdmin": "tailwindcss -i ./app/assets/stylesheets/active_admin.css -o ./app/assets/builds/active_admin.css --minify -c tailwind-active_admin.config.js"
},
```

Add a rake task based on the [cssbundling-rails](https://github.com/rails/cssbundling-rails/blob/123569431c0e303c1aca8b275b2d1bc3c68a3a63/lib/tasks/cssbundling/build.rake#L65) build task:

```ruby
# lib/tasks/build.rake
# change npm to whatever you're using

namespace :css do
  task :build_admin do
    system("npm run build:cssAdmin")
  end
end

if Rake::Task.task_defined?("css:build")
  Rake::Task["css:build"].enhance([ "css:build_admin" ])
end
```

I enhanced `css:build` but `assets:precompile` would work as well.

If using Procfile.dev for development add the extra script:

```bash
web: env RUBY_DEBUG_OPEN=true bin/rails server
js: npm run build --watch
css: npm run build:css --watch
css: npm run build:cssAdmin --watch
```

I haven't tested this in production yet but running `rails assets:precompile` works as expected adding all files to `public/assets/`

This was tested using `cssbundling-rails` but Im guessing a similar approach could be used with tailwind executable installed by `tailwindcss-rails`.
